Convert E_ERROR to thrown Error in extensions

trowski opened the pull request:
This PR converts most of the fatal errors in extensions to thrown Errors: [Uses of `php_error_docref()` with `E_ERROR`](http://lxr.php.net/search?&q=%22php_error_docref%28NULL%2C+E_ERROR%22&project=PHP_TRUNK&sort=fullpath).

Note that not all results shown in the link above were converted to thrown Errors. Those dealing with memory allocation, failing to register a class, or other unrecoverable errors have remained fatal. I also did not covert any fatal errors in the Soap extension. The extension appears to rely heavily on bailout behavior, so fatals could not easily be converted to thrown exceptions. Perhaps someone with more experience with the Soap extension would be able to do so.

I would like some more people to review the changes I've made before merging to ensure memory is being freed where needed and that I'm exiting each function appropriately after throwing the Error.

nikic commented on ext/date/php_date.c:
```diff
@@ -5015,7 +5018,8 @@ static zval *date_period_read_property(zval *object, zval *member, int type, voi
 	if (type != BP_VAR_IS && type != BP_VAR_R) {
-		php_error_docref(NULL, E_ERROR, "Retrieval of DatePeriod properties for modification is unsupported");
+		zend_throw_error(zend_ce_error, "Retrieval of DatePeriod properties for modification is unsupported");
+		return NULL;
```
This should probably return `&EG(uninitialized_zval)` instead.

nikic commented on ext/date/php_date.c:
```diff
@@ -3699,7 +3700,8 @@ PHP_METHOD(DateTimeZone, __set_state)
 	if(php_date_timezone_initialize_from_hash(&return_value, &tzobj, myht) != SUCCESS) {
-		php_error_docref(NULL, E_ERROR, "Timezone initialization failed");
+		zend_throw_error(zend_ce_error, "Timezone initialization failed");
+		RETURN_FALSE;
```
`return_value` was initialized to a `date_ce_timezone` object above, so this likely needs to be dtor'd here.

nikic replied:
Off-topic: The `return_value` arg of the `php_date_timezone_initialize_from_hash` function appears to no longer be used.

nikic commented on ext/dom/document.c:
```diff
@@ -2223,11 +2223,13 @@ PHP_METHOD(domdocument, registerNodeClass)
 		DOM_GET_OBJ(docp, id, xmlDocPtr, intern);
 
 		if (dom_set_doc_classmap(intern->document, basece, ce) == FAILURE) {
```
Looking at http://lxr.php.net/xref/PHP_TRUNK/ext/dom/php_dom.c#209 this seems to be simply dead code. We should adjust the function to be void instead.

nikic commented on ext/intl/transliterator/transliterator_class.c:
```diff
@@ -183,7 +183,7 @@ static zend_object *Transliterator_clone_obj( zval *object )
 
 			err_msg = intl_error_get_message( TRANSLITERATOR_ERROR_P( to_orig ) );
-			php_error_docref( NULL, E_ERROR, "%s", ZSTR_VAL(err_msg) );
+			zend_throw_error( zend_ce_error, "%s", ZSTR_VAL(err_msg) );
```
Not sure about this, but I don't think this properly destroys `ret_val` (it only destroys the `utrans` it holds).

trowski replied:
There's a comment below about not destroying `tempz`, which is derived from `ret_val`, so I'm not entirely sure if either should be freed.

nikic replied:
You're right, `ret_val` has to stay alive here. In which case the `transliterator_object_destroy` is redundant, though probably doesn't hurt either.

nikic commented on ext/ldap/ldap.c:
```diff
@@ -1822,8 +1822,9 @@ PHP_FUNCTION(ldap_modify_batch)
-				RETURN_FALSE;
+				zend_throw_error(zend_ce_error, "Unknown and uncaught modification type.");
+				RETVAL_FALSE;
+				goto cleanup;
```
At this point `ldap_mods` will be partially initialized, but `cleanup` assumes a fully initialized array, so this will likely try to free random memory.

trowski replied:
Could this be fixed by adding `if (ldap_mods[i] == NULL) break;` as the first line in the for-loop in the cleanup block?

nikic replied:
In that case you'll also have to `memset` it before the loop.

trowski replied:
Fixed by calling `efree(ldap_mods[i])` and setting `num_mods` to `i` before going to `cleanup`.

nikic commented on ext/mbstring/php_mbregex.c:
```diff
@@ -943,8 +944,9 @@ static void _php_mb_regex_ereg_replace_exec(INTERNAL_FUNCTION_PARAMETERS, OnigOp
-					/* zend_error() does not return in this case */
+					zend_throw_error(zend_ce_error, "Failed evaluating code: %s%s", PHP_EOL, ZSTR_VAL(eval_buf.s));
+					onig_region_free(regs, 0);
+					RETVAL_EMPTY_STRING();
```
After this, the `onig_region_free` in L1007 will still be reached, resulting in a double free.

nikic commented on ext/pdo_odbc/pdo_odbc.c:
```diff
@@ -123,7 +124,7 @@ PHP_MINIT_FUNCTION(pdo_odbc)
 		pdo_odbc_pool_on = SQL_CP_OFF;
 	} else {
-		php_error_docref(NULL, E_ERROR, "Error in pdo_odbc.connection_pooling configuration.  Value MUST be one of 'strict', 'relaxed' or 'off'");
+		zend_throw_error(zend_ce_error, "Error in pdo_odbc.connection_pooling configuration.  Value MUST be one of 'strict', 'relaxed' or 'off'");
```
This is in MINIT, we should not throw exceptions here. Instead this should probably be an `E_CORE_ERROR`.

trowski replied:
This was another I wondered about changing to an exception. I'll change it to an E_CORE_ERROR then.

nikic commented on ext/session/session.c:
```diff
@@ -480,7 +481,7 @@ static void php_session_initialize(void) /* {{{ */
 
 	if (!PS(mod)) {
-		php_error_docref(NULL, E_ERROR, "No storage module chosen - failed to initialize session");
+		zend_throw_error(zend_ce_error, "No storage module chosen - failed to initialize session");
```
Did you verify that we can gracefully handle a session initialization failure?

nikic commented on ext/session/tests/bug60860.phpt:
```diff
@@ -12,6 +12,12 @@ echo "ok\n";
+Warning: session_start(): user session functions not defined in %s on line %d
 
-Fatal error: session_start(): Failed to initialize storage module: user (path:%s) in %s on line 3
+Warning: session_start(): Cannot send session cache limiter - headers already sent (output started at %s:%d) in %s on line %d
```
An additional warning being thrown is usually an indication for code not aborting sufficiently early in response to an exception.

trowski replied:
Ok, I wondered about this. Failing to initialize the session storage should probably just be a fatal error anyway, so I'll revert this change.

nikic commented on ext/simplexml/simplexml.c:
```diff
@@ -250,7 +250,7 @@ static zval *sxe_prop_dim_read(zval *object, zval *member, zend_bool elements, z
 			/* This happens when the user did: $sxe[]->foo = $value */
-			php_error_docref(NULL, E_ERROR, "Cannot create unnamed attribute");
+			zend_throw_error(zend_ce_error, "Cannot create unnamed attribute");
 			return NULL;
```
Should return `&EG(uninitialized_zval)` here as well (and other occurrence in this function).

nikic commented on ext/simplexml/simplexml.c:
```diff
@@ -277,7 +277,7 @@ static zval *sxe_prop_dim_read(zval *object, zval *member, zend_bool elements, z
 		    node->parent->type == XML_DOCUMENT_NODE) {
 			/* This happens when the user did: $sxe[]->foo = $value */
-			php_error_docref(NULL, E_ERROR, "Cannot create unnamed attribute");
+			zend_throw_error(zend_ce_error, "Cannot create unnamed attribute");
```
For reviewers: As this is behind a `!member` guard the string copy in L259 couldn't have happened, so we don't have to free it.

nikic commented on ext/simplexml/simplexml.c:
```diff
@@ -571,7 +571,7 @@ static int sxe_prop_dim_write(zval *object, zval *member, zval *value, zend_bool
 			if (!member || Z_TYPE_P(member) == IS_LONG) {
 				if (node->type == XML_ATTRIBUTE_NODE) {
-					php_error_docref(NULL, E_ERROR, "Cannot create duplicate attribute");
+					zend_throw_error(zend_ce_error, "Cannot create duplicate attribute");
```
A copy of `value` might have been made in L520, so that may have to be freed here?

trowski replied:
Added the same check and free below the throw for `value` that is at the end of the function.

nikic commented on ext/spl/spl_array.c:
```diff
@@ -787,7 +787,10 @@ static HashTable *spl_array_get_properties(zval *object) /* {{{ */
 	if (intern->nApplyCount > 1) {
-		php_error_docref(NULL, E_ERROR, "Nesting level too deep - recursive dependency?");
+		zend_throw_error(zend_ce_error, "Nesting level too deep - recursive dependency?");
+		ALLOC_HASHTABLE(result);
```
This hashtable will leak. Also not sure if this exception will be handled gracefully by callers.

nikic commented on ext/spl/spl_directory.c:
```diff
@@ -196,7 +196,7 @@ static inline void spl_filesystem_object_get_file_name(spl_filesystem_object *in
 		case SPL_FS_INFO:
 		case SPL_FS_FILE:
 			if (!intern->file_name) {
-				php_error_docref(NULL, E_ERROR, "Object not initialized");
```
This function seems to be used in a bunch of places. Did you check that they can all handle the exception?

trowski replied:
No, making this one an exception was a mistake, I will revert it.

nikic commented on ext/spl/spl_iterators.c:
```diff
@@ -1392,7 +1392,7 @@ int spl_dual_it_call_method(char *method, INTERNAL_FUNCTION_PARAMETERS)
 	ZVAL_STRING(&func, method, 0);
 	if (!zend_is_callable(&func, 0, &method)) {
-		php_error_docref(NULL, E_ERROR, "Method %s::%s() does not exist", intern->inner.ce->name, method);
+		zend_throw_error(zend_ce_error, "Method %s::%s() does not exist", intern->inner.ce->name, method);
```
OT: This code is dead, we should drop it.

trowski commented:
I've addressed the issues raised by @nikic (excluding the one in `ext/intl/transliterator/transliterator_class.c` since I'm also unsure of what to do there).

@weltling and @KalleZ perhaps you'd like to review it as well. I was hoping to merge this for beta 1.

weltling commented:
@trowski can't compile it right now, 

`ext\ldap\ldap.c(1825): error C2065: 'zend_ce_error': undeclared identifier`

Thanks

weltling commented:
@trowski, from the test side it looks good. Though please add the missing header to ext/ldap and do a full build before merging.

From the conversion side - yeah, there are several places that need to be done extra carefully. Fe I remember in ext/date there was some functionality in wakeup that explicitly depends on a fatal error.

 Thanks.

KalleZ commented:
Just ccing @derickr here for the DateTime stuff

KalleZ commented:
@trowski Me and @weltling have talked this over, and we think that it is a little too late to change extensions to use exceptions, the RFC[1] by @nikic does not mention extensions anywhere, although it would make sense to do so. I understand this is a "mere" consistency improvement, but I think we should postpone this to 7.1, mainly because it is a little late in the release process (Beta 1 being tagged tomorrow, marking the feature freeze). By postponing this to 7.1, we can also cover more extensions and cases in a clear RFC.

[1]  https://wiki.php.net/rfc/engine_exceptions

trowski commented:
@weltling Thanks for trying to compile the ldap extension and noticing the overlooked include. I'm on OS X and had the same problem building ldap as being discussed on the mailing list, so I was unable to build it myself.

@KalleZ Waiting until 7.1 sounds reasonable. I'm sorry I was unable to do this sooner. I'll just leave this PR open and we can revisit it after 7.0 has been branched. Then we can do a more thorough examination and perhaps look at the Soap extension as well.
